### PR TITLE
Initialize Adagrad parameter state lazily in the C++ optimizer

### LINKED ISSUE #87706: C++ Adagrad optimizer doesn't initialize parameter state
### 🐛 Describe the bug

optim/adagrad.cpp lacks code to initialize parameter state.

Running this code:

```
 auto x=torch::tensor({0.5, 2.0, 4.0}, torch::requires_grad());
 auto y=torch::tensor({1.0, 2.0, 3.0});
 auto o=torch::optim::Adagrad(std::vector<torch::Tensor>{}, torch::optim::AdagradOptions(0.1));
 auto& p=o.param_groups();
 p[0].params().push_back(x);
 auto l=torch::nn::functional::mse_loss(x,y);
 l.backward();
 o.step();
```
will throw:
_'state_[c10::guts::to_string(p.unsafeGetTensorImpl())] != nullptr INTERNAL ASSERT FAILED at "../torch/csrc/api/src/optim/adagrad.cpp":80, please report a bug to PyTorch. state found NULL for the Tensor_ 

the other optimizers have code that searches for the state and initializes if not found, e.g. from sgd.cpp:

```
        auto param_state = state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
        if(param_state == state_.end()) {
          buf = torch::clone(d_p).detach();
          auto state = std::make_unique<SGDParamState>();
          state->momentum_buffer(buf);
          state_[c10::guts::to_string(p.unsafeGetTensorImpl())] = std::move(state);
        } else {
          buf = static_cast<SGDParamState&>(*param_state->second).momentum_buffer();
          buf.mul_(momentum).add_(d_p, 1 - dampening);
        }
```
adagrad.cpp doesn't contain this kind of code.

### Versions

This error in all versions since PR  #29335

### PR DESCRIPTION
**Pull request by egeboy35**
Fixes #87706.

`torch::optim::Adagrad` created per-parameter state only in its constructor, so a
parameter added afterwards — via `add_param_group`, or by pushing onto an existing
group's `params()` as in the issue — reached `step()` with no state and tripped
`TORCH_INTERNAL_ASSERT(... "state found NULL for the Tensor ")`.

Adagrad was the only C++ optimizer doing this; `sgd.cpp:82`, `adam.cpp:82`,
`adamw.cpp:82`, `rmsprop.cpp:84` and `lbfgs.cpp:422` all create state on first use inline
in `step()`. There is no `_init_group` equivalent on the C++ side, so this adds the same
inline branch rather than a new pattern.

The constructor's eager loop stays. `adagrad.py` initializes both eagerly in `__init__`
and lazily in `_init_group`, and `OptimTest.OptimizerAccessors` depends on the eager half
twice: it reads `optimizer.state()` without calling `step()`, and it expects
`add_param_group` to reject a duplicate parameter — a check implemented as
`state_.count(...) == 0` (`optimizer.cpp:300-304`), which silently accepts duplicates if
state only appears after the first `step()`. #106896 deleted that loop and went red on ten
jobs. Both sites now share `Adagrad::make_param_state`, seeded from the optimizer defaults
to match the constructor and `adagrad.py:172`.

Complex is out of scope: `Adagrad::step` has no `view_as_real` path, so C++ Adagrad is
already wrong for complex regardless of state init. Same for Adam/AdamW/RMSprop.

### Test

`OptimTest.AddParameter_Adagrad` covers both late-arrival paths and gives the first group
an `initial_accumulator_value` differing from the defaults, so it pins which one the lazy
path reads.

I can now run the gtest binaries locally, so here is the test run against both versions of
the library in one CPU-only Windows build tree. First with `adagrad.cpp` and `adagrad.h`
reverted to their state on this PR's merge base (`7744b78e00e`) and `test_api` rebuilt —
that is, the test against unmodified main. Absolute paths are shortened in both blocks:

```
$ ./test_api.exe --gtest_filter=OptimTest.*Adagrad*:OptimTest.OptimizerAccessors

[ RUN      ] OptimTest.AddParameter_Adagrad
test/cpp/api/optim.cpp(500): error: Expected: optimizer.step() doesn't throw an exception.
  Actual: it throws c10::Error with description "state_[p.unsafeGetTensorImpl()] != nullptr
  INTERNAL ASSERT FAILED at "torch/csrc/api/src/optim/adagrad.cpp":78, please report a bug
  to PyTorch. state found NULL for the Tensor ...
[  FAILED  ] OptimTest.AddParameter_Adagrad (607 ms)

[==========] 7 tests from 1 test suite ran. (3295 ms total)
[  PASSED  ] 6 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] OptimTest.AddParameter_Adagrad
```

The other six stay green, `OptimTest.OptimizerAccessors` among them — that is the test the
constructor's eager loop exists for, so this also shows the revert isolates the new
behaviour rather than disturbing the old one.

With the fix restored and `test_api` rebuilt from the same tree:

```
$ ./test_api.exe --gtest_filter=OptimTest.*Adagrad*:OptimTest.OptimizerAccessors

[       OK ] OptimTest.OptimizerAccessors (2 ms)
[       OK ] OptimTest.XORConvergence_Adagrad (1964 ms)
[       OK ] OptimTest.ProducesPyTorchValues_Adagrad (502 ms)
[       OK ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecay (501 ms)
[       OK ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecayAndLRDecay (581 ms)
[       OK ] OptimTest.AddParameter_Adagrad (0 ms)
[       OK ] OptimTest.MergeWithDefaultOptions_Adagrad (0 ms)
[  PASSED  ] 7 tests.
```

It also passed in CI on `74042392cc4`, the last commit CI has been approved for on this
PR; the current head `9131265e04a` only trims comments in the test body and its runs are
still awaiting fork approval. From
`test-reports/cpp-unittest/test_libtorch/test_api.xml` in
`linux-jammy-py3.10-clang21-asan / test (default, 1, 18)`:

```
AllTests:  tests=1023  failures=0  errors=0
OptimTest: tests=45    failures=0  errors=0
  AddParameter_Adagrad   status=run result=completed failures=0
  OptimizerAccessors     status=run result=completed failures=0
```

Developed with AI assistance (Claude); I reviewed the change and the reasoning above.

### CODE REVIEW

**Comment by janeyx99** [comment]
@claude please review this PR with the review skill. Is this the right approach to the task? My initial read is that we should align with whatever the other C++ optims are doing; if they have a separate logic like _init_group where initialization happens, then it'd be good to be parallel to those for consistency.

File: torch/csrc/api/src/optim/adagrad.cpp [hunk 1]
```diff
@@ -72,10 +82,14 @@ Tensor Adagrad::step(LossClosure closure) {
...
-      TORCH_INTERNAL_ASSERT(
-          state_[p.unsafeGetTensorImpl()] != nullptr,
-          "state found NULL for the Tensor ",
-          p);
+      auto param_state = state_.find(p.unsafeGetTensorImpl());
+      auto& defaults = static_cast<AdagradOptions&>(*defaults_);
+
+      // State initialization
+      if (param_state == state_.end()) {
+        state_[p.unsafeGetTensorImpl()] =
```
Comment (Skylion007) [inline torch/csrc/api/src/optim/adagrad.cpp]: Slight nit: should be a try_emplace or
 emplace right?

**Comment by janeyx99** [comment]
Please address the code quality concerns in the claude review. Were you able to run the test locally/can you paste the output in the PR body?

File: torch/csrc/api/src/optim/adagrad.cpp (hunk 1 above)
Comment (egeboy35) [reply torch/csrc/api/src/optim/adagrad.cpp]: Done in 70625fa059c — switched to `emplace` and reused the iterator it returns, which also drops the third hash lookup (the block is now one `find` plus one `emplace`, and no longer calls `operator[]` at all). Worth noting `ska::flat_hash_map` has no `try_emplace`, only `emplace`. This does leave Adagrad differing from `sgd.cpp:87` / `adam.cpp:97` / `adamw.cpp:102` / `rmsprop.cpp:98`, which still assign through `operator[]` — happy to align those in a follow-up if you'd like, I kept it out of the bugfix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by egeboy35** [comment]
Thanks both. On the consistency question: there is no `_init_group` equivalent in the
C++ frontend — all five optimizers create their state inline in `step()`, guarded by
`state_.find(...) == state_.end()` (`sgd.cpp:82`, `adam.cpp:82`, `adamw.cpp:82`,
`rmsprop.cpp:84`, `lbfgs.cpp:422`). Adagrad was the only one building state solely in
its constructor, which is exactly why a parameter added later hit the assert. This PR
adds the same inline branch in the same place, so it is now parallel to the other five
rather than introducing a new pattern.

Pushed 3a7545f3383 and 70625fa059c addressing the review:

- **Deduped the seeding.** The constructor and the lazy path now both call a single
  `Adagrad::make_param_state`, so "same as the constructor" is structural instead of a
  comment that can silently go stale. That also folds in the `p` vs `p.data()` and
  unqualified `MemoryFormat::Preserve` fixes.
- **Hoisted the `find` result** into `param_state` and tested it against `state_.end()`,
  matching the siblings line for line.
- **Widened the test**: both late arrivals — the `push_back` one and the
  `add_param_group` one — are now checked for `step == 1` and the expected `sum`, not
  just `allclose`.
- **Commit message** now records that the eager constructor loop also keeps
  `OptimTest.OptimizerAccessors`' duplicate-parameter assertion alive, not only its
  `state()` read.
- **Switched the insert to `emplace`** per @Skylion007's nit, reusing the returned
  iterator so the block is one `find` plus one `emplace` and never calls `operator[]`.
  (`ska::flat_hash_map` has no `try_emplace`.) This leaves Adagrad differing from the
  four siblings, which still assign through `operator[]` — glad to align them in a
  follow-up.

Two things deliberately left out:

- **Making the seed group-aware.** `initial_accumulator_value` is a per-group
  `TORCH_ARG`, but the constructor seeds every group from `defaults` and
  `adagrad.py:172` does the same in `_init_group`, so honouring a per-group value would
  be a behaviour change in both paths. Happy to file it as a follow-up — it felt wrong
  to fold into a bugfix.
- **The present-but-null hole.** The old `operator[]`-based assert also caught a
  default-inserted null entry; `find(...) == end()` does not. Every sibling has the same
  gap, so I matched them rather than diverging, but flagging it so the trade is on the
  record rather than incidental.

Complex is still out of scope: `Adagrad::step` has no `view_as_real` path (contrast
`adagrad.py:417-422`), so C++ Adagrad is already numerically wrong for complex however
the state is initialized — changing only the seed would be a half-fix. Glad to open a
separate issue for complex support across the C++ optimizers if that is useful.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by janeyx99** [comment]
@claude review the updates

**Comment by janeyx99** [comment]
@egeboy35 would still love to see the test output in the PR body

**Comment by janeyx99** [comment]
Address the nits + do update the PR body to be reflective of the final change.

**Comment by janeyx99** [comment]
@egeboy35 I think you can make the PR description way less verbose than AI and just as informative.

I'm not sure this test was run in CI and passes, so I'm hesitant to approve this PR until it's proven to pass.

File: test/cpp/api/optim.cpp [hunk 2]
```diff
@@ -470,6 +470,57 @@ TEST(OptimTest, AddParameter_LBFGS) {
   // REQUIRE this doesn't throw
 }
 
+TEST(OptimTest, AddParameter_Adagrad) {
```
Comment (janeyx99) [inline test/cpp/api/optim.cpp:473]: for each comment in this test, please consider whether it is truly helpful and understandable. If not, please remove.

**Review by janeyx99 — CHANGES_REQUESTED** [review]

**Comment by egeboy35** [comment]
@janeyx99 the nits are in `74042392cc4` and I have rewritten the PR body to describe the
final shape rather than the original one.

On the test output: I owe you an honest answer rather than a paste. I cannot run the gtest
binary here — this machine's Device Guard policy refuses to launch freshly linked
executables:

```
build/bin/c10_ArrayRef_test.exe was blocked by your organization's Device Guard policy.
```

DLLs do load, so I ran the test's exact scenario through a small Python extension module
against a `torch_cpu` built *without* this fix, and it reproduces the reported failure:

```
step()                        : THREW
  state_[p.unsafeGetTensorImpl()] != nullptr INTERNAL ASSERT FAILED at
  "torch/csrc/api/src/optim/adagrad.cpp":78, please report a bug to PyTorch.
  state found NULL for the Tensor ...
```

That is the half I think matters most, since it is what #106896's final test could no
longer demonstrate: the scenario the new test sets up really is the one that trips the
assert today, so the test is not vacuous.

For the passing half I am relying on CI. The run on `70625fa059c` was green across all 156
checks, including `linux-jammy-py3.10-gcc11 / test (default, 1, 10)`, which
`.ci/pytorch/test.sh` routes through `test_libtorch 1` to `test_libtorch_api`. The current
head only differs from it by the three cosmetic nits, and its workflows are still waiting
on approval to run.

Happy to paste the literal gtest lines if you would rather wait for that run — just say so
and I will follow up once it completes.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by janeyx99** [comment]
Not all my review comments are addressed. and the PR body is still incredibly long.

**Comment by egeboy35** [comment]
@janeyx99 fair on both counts — I have cut the description down considerably.

On the test: it does run in CI and it passes. From
`test-reports/cpp-unittest/test_libtorch/test_api.xml` in the
`linux-jammy-py3.10-clang21-asan / test (default, 1, 18)` job of the green run on
`74042392cc4`:

```
AllTests:  tests=1023  failures=0  errors=0
OptimTest: tests=45    failures=0  errors=0
  AddParameter_Adagrad   status=run result=completed failures=0
  OptimizerAccessors     status=run result=completed failures=0
```

`OptimizerAccessors` passing is the other half worth noting — it is the test #106896 broke
by removing the constructor's eager init, so it confirms keeping that loop was the right
call.

Sorry for the earlier wall of text.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/cpp/api/optim.cpp (hunk 2 above)
Comment (egeboy35) [reply test/cpp/api/optim.cpp:473]: Done in 9131265e04a. I dropped the paragraph restating that both late arrivals are seeded from the optimizer defaults — the comment above the param group already says the assertions pin which value is used, and the rationale belongs in the PR rather than the test — and shortened the opening comment and the `at()` note.

What I kept, and why: why the test exists at all; why the group's `initial_accumulator_value` and `lr` are set the way they are (simplifying either one silently decouples the test from what it checks); `at()` rather than `operator[]`, since `operator[]` would default-insert null and segfault instead of failing; and where the expected `sum` of 1.5 comes from. Happy to cut further if any of those still read as noise.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by egeboy35** [comment]
@janeyx99 everything you asked for is in: the nits are addressed in `74042392cc4`, the test
comments trimmed in `9131265e04a`, and the description rewritten and cut down.

On the test actually passing — from `test-reports/cpp-unittest/test_libtorch/test_api.xml` in
the green `linux-jammy-py3.10-clang21-asan / test (default, 1, 18)` run:

```
OptimTest: tests=45  failures=0  errors=0
  AddParameter_Adagrad   status=run result=completed failures=0
  OptimizerAccessors     status=run result=completed failures=0
```

The review is still marked as requesting changes, which holds the PR. Would you mind taking
another look when you get a moment? Happy to keep iterating if anything still reads wrong.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by egeboy35** [comment]
@janeyx99 the comment trimming you asked for landed on the branch as `9131265e04a` on Aug 26 — it cuts the comments in `AddParameter_Adagrad` from 14 lines to 7, dropping the paragraph that restated what the code already says and keeping only what the code cannot express.

The `CHANGES_REQUESTED` is still standing, so the PR is blocked even though the request is addressed. Could you re-review, or dismiss it if you would rather someone else pick this up?

_🤖 Drafted with [Claude Code](https://claude.com/claude-code)_

**Comment by egeboy35** [comment]
@janeyx99 gentle ping - both review points have been addressed and this has been sitting since Aug 26.

- Your note on the test comments: done in 9131265e04a. I dropped the paragraph restating that both late arrivals are seeded from the optimizer defaults, and shortened the opening comment and the `at()` note. What I kept is why the test exists and why the group's `initial_acc` matters, since neither is inferable from the assertions.
- @Skylion007's `emplace` nit: done in 70625fa059c - one `find` plus one `emplace` reusing the returned iterator, no `operator[]`.

CI is green and the PR is mergeable. Happy to make further changes if the comments still read as noise to you - no rush, just flagging that it is unblocked on my side.
